odbc/src/api/functions.rs

```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // TINYINT bits from FULL_MONGOSQL_CONVERSION_TARGET_MASK via
+            // bitwise XOR.
+            const TIMESTAMP_BITMASK: u32 = FULL_MONGOSQL_CONVERSION_TARGET_MASK
+                ^ SQL_CVT_INTEGER
+                ^ SQL_CVT_SMALLINT
+                ^ SQL_CVT_TINYINT;
+            return i16_len::set_output_fixed_data(
+                &TIMESTAMP_BITMASK,
+                info_value_ptr,
+                string_length_ptr,
+            );
+        }
+        // SQL_CONVERT_BINARY
+        InfoType::ConvertBinary
+        // SQL_CONVERT_DATE
+        | InfoType::ConvertDate
+        // SQL_CONVERT_TIME
+        | InfoType::ConvertTime
+        // SQL_CONVERT_VARBINARY
+        | InfoType::ConvertVarBinary
+        // SQL_CONVERT_LONGVARBINARY
+        | InfoType::ConvertLongVarBinary
+        // SQL_CONVERT_GUID
+        | InfoType::ConvertGuid => {
+            // Binary, Date, Time, VarBinary, LongVarBinary, and GUID
+            // are invalid CAST source types.
+            return i16_len::set_output_fixed_data(&SQL_U32_ZERO, info_value_ptr, string_length_ptr);
+        }
+        // SQL_GETDATA_EXTENSIONS
+        InfoType::GetDataExtensions => {}
+        // SQL_COLUMN_ALIAS
+        InfoType::ColumnAlias => {
+            // MongoSQL does support column aliases.
+            return i16_len::set_output_wstring(
+                SQL_INFO_Y,
+                info_value_ptr as *mut WChar,
+                buffer_length as usize,
+                string_length_ptr,
+            );
+        }
+        // SQL_GROUP_BY
+        InfoType::GroupBy => {}
+        // SQL_ORDER_BY_COLUMNS_IN_SELECT
+        InfoType::OrderByColumnsInSelect => {
+            // MongoSQL does require ORDER BY columns to be in the SELECT list.
+            return i16_len::set_output_wstring(
+                SQL_INFO_Y,
+                info_value_ptr as *mut WChar,
+                buffer_length as usize,
+                string_length_ptr,
+            );
+        }
+        // SQL_OWNER_USAGE
+        InfoType::OwnerUsage => {
+            // SQL_OWNER_USAGE is replaced by SQL_SCHEMA_USAGE in newer
+            // ODBC versions. They use the same numeric value.
+            //
+            // MongoSQL does not have a concept of a "schema" which is a
+            // level above "catalog" (or "database") in the data hierarchy.
+            return i16_len::set_output_fixed_data(&SQL_U32_ZERO, info_value_ptr, string_length_ptr);
+        }
+        // SQL_CATALOG_USAGE
+        InfoType::CatalogUsage => {
+            // TODO: (code review) there is no way to indicate this is read-only. The entire DML bitmask
+            //       must be used to indicate we support SELECT. However, that also includes INSERT,
+            //       UPDATE, and DELETE.
+            return i16_len::set_output_fixed_data(&SQL_CU_DML_STATEMENTS, info_value_ptr, string_length_ptr);
+        }
+        // SQL_SPECIAL_CHARACTERS
+        InfoType::SpecialCharacters => {
```

Comment:
The important thing here is **when do we require the identifier to be a quoted identifier.**
Postgres returns only "_".
I'd say that we need the identifier to be quoted if it contains `, ", ', . or $ (the last 2 are not supported yet), no?
Any other character that you see would throw the parser or algebrizer off?  I so, add it to this list.

Reply:
The [spec says](https://docs.google.com/document/d/14DnyMSzVNRkoyib8yM62MPJIrkrG1pE0KjRYOWOUuDc/edit#heading=h.ea137t3xw581) that regular (unquoted) identifiers can only consist of ascii letters, digits, and underscore, and everything else must be quoted:
```
<regular identifier> ::= ([A-Za-z] | "_")[A-Za-z0-9_]*

<delimited identifier> ::= " <identifier character>* "
                         | ` <identifier character>* `

<identifier character> ::= [^\x00]
```
Plus we have spec tests asserting this behavior, so to match your suggestion would require a spec change. We can go down that route, but we may want to discuss this as a team before committing to this.

Reply:
umh, I'm not sure why you'd require a spec change?
It seems more like this info type has a very bad format and it is not realistic to return a string with all the possible character that we accept. Returning nothing does not seem that great either, because it implies that we don't need the quotes for any characters.
I would love to be able to provide a regex but if application expects a list of chars to know if they need to quote identifier or not, then it won't work.
Covering the basic most used chars is maybe not 100% better because then it is incorrect too, but it seems more usable.
Ultimately, I believe most app always put identifier in quotes to be on the safe side.
Let's talk about this a bit more tomorrow

Reply:
Updated the implementation to return a string containing some special characters like you brought up. lmk what you think!

Reply:
I think that's already quite a lot and will probably go a long way 👍 Thank you for taking the time to think about useful ones :)